### ISSUE TITLE
Add support for manual resizing

### ISSUE THREAD

**Issue by flack** [codemirror/codemirror5#850]
Sorry if this is a duplicate, but a quick search of the issue list found nothing that seemed related.

I was wondering if CodeMirror could be made to support the CSS resize attribute. So that I could write something like this:

``` css
.CodeMirror-scroll
{
    resize: vertical;
}
```

In the current release (2.34), this actually renders a resize handle, but I can't move it. Or am I just doing it wrong?

**Comment by mindplay-dk** [comment]
I'd like to just have a simple drag-handle, a bar running along the bottom edge of the editor, that the user can drag to resize vertically.

This seems like a pretty basic requirement? A standard textarea can be made user-resizeable.

Yet, I can't find an add-on or plugin or fiddle or anything else demonstrating this.

Has anyone figured out how to make the window resizable?

(it would be really nice to have a standard add-on listed for this basic feature - many of the listed/standard add-ons are far more advanced features than this...)

**Comment by marijnh** [comment]
The reason this isn't an addon is probably that it's rather easy to implement, and since it involves UI elements, providing a general-enough implementation that fits everyone's needs is more involved than just coding one-off implementations per application.

Given that we're not adding new addons to the main distribution anymore, I'm going to close this as out of scope. But if you want to tackle it in a reusable way and publish it on NPM, that'd be wonderful.

**Comment by mindplay-dk** [comment]
I disagree that anything to do with drag-and-drop in DOM/JS is easy to implement.

It took me an hour just to get [this far](https://jsfiddle.net/mindplay/rs2L2vtb/) and I still have no idea how to make it capture the mouse, so that the drag operation continues if you drag outside the window/frame. But I'm probably doing something wrong. There's a million ways drag-and-drop goes horribly wrong with DOM and JS.

IMO the lib should offer functionality on par with a plain textarea. The question gets asked enough on StackOverflow, and nobody gives any practical answer or an example, just short remarks like "do it yourself", or in one case how to do it with jQuery UI, which isn't going to be the answer most people are looking for in this day and age ;-)

**Comment by mindplay-dk** [comment]
If somebody else needs it (or wants to take point on turning it into an add-on) here's a mostly working [vertical resize bar](https://jsfiddle.net/mindplay/rs2L2vtb/). If you click and drag repeatedly, it still sometimes gets confused and "sticks", but most of the time it works. Sorry, I did my best ;-)

**Comment by Sphinxxxx** [comment]
@mindplay-dk https://github.com/Sphinxxxx/cm-resize

**Comment by lunakid** [comment]
Joining the club with probably the simplest (and therefore perhaps least brittle) implementation I've found so far: [CodePen](https://codepen.io/sakifargo/pen/KodNyR), or [StackOverflow](https://stackoverflow.com/a/49247857/1479945) (with explanations). *(Proof-of-concept example impl. only!)*

While functionally OK (on browsers supporting CSS `resize`, so no Edge yet), the appearance is not perfect:  there's some extra bottom padding to make room for the resize handle on e.g. Chrome.

The reason is because (a) CM, unaware of the handle, happily draws the scrollbars, the filler, and the text box there -- which (b) on WebKit includes the scrollbars in the rect size, and (c) I couldn't find a way to get the handle still rendered visible (on top of the other items + also clickable) on Chrome.

Firefox, OTOH, draws the handle on the top of the other bits -- even over scrollbar arrows, if there happens to be one there --, and you can grab it, too, so resizing works even without any extra empty space.

However, even for FF, just for a tidy appearance, it would be desirable if CM could either detect, or be instructed, that there's a resize handle, so room should be made for it when sizing the scrollbars, or drawing the filler. (I guess it's problematic because of the browser-depoendency, i.e. the WebKit sizing discrepancy regarding scrollbars. Also, with proper CSS support this should be mostly implicit.)

Being no expert in the matter, hopefully I'm missing something obvious, which could solve this easily, so CM could finally have a resize handle without resorting to hacky workarounds.

To recap, CM should simply:

* hook the appropriate Observer object to get the resize events,
* and make sure the CSS resize handle can appear.

@marijnh: as this approach does *not* add any custom UI to CM, could it be implemented right in the lib (also eliminating the extra DIV), to enable sizing merely via the CSS `resize`? That would be wonderful.

*(BTW: I've also tried styling the `.CodeMirror` DIV directly with `resize: both`, and hooking the resize events (via the same `...Observer` handlers as in the linked example), but I failed to get the scrollbars right with that approach, so gave up and resorted to an outer container div with the `resize` prop.)*

**Comment by JoseTadeuMarques** [comment]
in case this will help anyone, this did the trick for me:

.CodeMirror {
      resize: vertical;
      overflow: auto !important;
    }

I got a resizable Codemirror Instance

**Comment by lunakid** [comment]
Jose, I'm puzzled. Can you show an example? Do you mean newer CM versions support it out-of-the-box? Which version did you use? Are you 100% sure that the CM instance *itself* is resizing, and you didn't just see a resizable box with scrollbars? (Which is, of course, not the same.) Thanks!

**Comment by Sphinxxxx** [comment]
A small update: My **cm-resize** is now on NPM:

https://www.npmjs.com/package/cm-resize

**Comment by JoseTadeuMarques** [comment]
@lunakid 

The question in this ticket was this:
"I was wondering if CodeMirror could be made to support the CSS resize attribute. So that I could write something like this:"

for this, I used what I said :
.CodeMirror {
  resize: vertical;
  overflow: auto !important;
}

to try this out, open this:
https://codemirror.net/demo/simplemode.html

then go to the Codemirror instance and inspect it. Then, in the div with class="CodeMirror", try the css I just wrote. Then you will see the resizing working. Scrollbars are also applied obviously.

I might have used the wrong words though, I can agree, as I'm not sure the instance itself is being resized. The box is however being resized, and the resize property is being applied properly, as asked in this ticket!

I'd stick to @Sphinxxxx suggestion though with the new npm module tho

**Comment by Sphinxxxx** [comment]
@JoseTadeuMarques - One problem: If the editor needs horizontal scrolling, you now get two scrollbars, and only one of them works:

https://jsfiddle.net/cepg4t95/

Any way to avoid that?

**Comment by JoseTadeuMarques** [comment]
@Sphinxxxx use cm-resize :)

**Comment by lunakid** [comment]
Guys, it's a bit too early for the celebration just yet. ;) 

@JoseTadeuMarques, unfortunately, what you see is not what you think: the CM instance is *not* actually resizing just by setting the CSS. What you describe and what you saw is *only a HTML DIV resizing* (which is quite unsurprising, adding a CSS rule for just that).

To understand what CM resizing *really* means, i.e. resizing the *actual CM Javascript object* (with all its geometry-related internal state), rather than just resizing a HTML box on the screen, see e.g. my demo at [CodePen](https://codepen.io/sakifargo/pen/KodNyR), where e.g. the line numbers + line wrapping clearly evidences how the editor properly recalculates its state with that method -- which was necessary exactly because just setting CSS `resize` does not cut it. (I.e. it will *not* pass resize events to the CM instance to get it recalculate its state!)

(For those who want to understand the mechanics, please see the details in my original comment above.)

**Comment by ytachi0026** [comment]
@lunakid i see what you are doing this, but if i am working on ReactJS, i am not sure how to adapt your code in the life cycle. Any clue?

**Comment by ytachi0026** [comment]
@JoseTadeuMarques i tried your package, but it couldn't get it work.

**Comment by lunakid** [comment]
> @lunakid i see what you are doing this, but if i am working on ReactJS, i am not sure how to adapt your code in the life cycle. Any clue?

I'm not familiar with ReactJS, unfortunately. But I suppose you should be able to get the JS and CSS code [from the demo ](https://codepen.io/sakifargo/pen/KodNyR), and just copy it to the somewhere in React (I assume React shouldn't make this particularly difficult).
